Add Findmlpack.cmake (#3872)
mlpack/mlpack: doc/embedded/crosscompile_example.md

in `target_link_libraries`.

```cmake
## Add your source files to SOURCES_FILES list
set(SOURCE_FILES main.cpp)
add_executable(RandomForest main.cpp ${SOURCES_FILES})

# If needed, add any additional include directories here.
target_include_directories(RandomForest PRIVATE
  ${MLPACK_INCLUDE_DIRS}
  # Add more include directories here...
)

# If your application needs to link against more than just mlpack's
# dependencies, be sure to list them here.
target_link_libraries(RandomForest PRIVATE -static
  ${MLPACK_LIBRARIES}
  # List additional dependencies to link against here.
)
```

#### Optimization and cross-compilation

If you are interested in adding specific compiler flags to optimize operations
on your hardware, you can either set `CMAKE_CXX_FLAGS` manually, or look at the
copy of `CMake/crosscompile-arch-flags.cmake` in your project, find the
appropriate `ARCH_NAME` section, and add the new compilation flags to that file.

Once you have added all the source files and the headers for your applications,
you can create your own build directory and build the software using `cmake`,
and then `make`. Your cmake command should be similar to the following:

```sh
cmake \

>     -ARCH_NAME=(Check below) \

Reviewer: ```suggestion
    -DARCH_NAME=(Check below) \
```

    -DCMAKE_CROSSCOMPILING=ON \
    -DCMAKE_TOOLCHAIN_FILE=../CMake/crosscompile-toolchain.cmake \
    -DTOOLCHAIN_PREFIX=(Check below) \
    -DCMAKE_SYSROOT=(Check below) \
```

> In order to fill the `TOOLCHAIN_PREFIX` or the `CMAKE_SYSROOT`, use
> [this table](supported_boards.md).

Reviewer: ```suggestion
In order to fill the `TOOLCHAIN_PREFIX` and `CMAKE_SYSROOT` variables, use
[this table](supported_boards.md).
```

## Crosscompile mlpack example for an embedded hardware

In this article, we explore how to add mlpack to a CMake project that cross-compiles
code to embedded hardware.  See also these related other guides, which may be useful

> to read before this one:

Reviewer: ```suggestion
to read before this one:

```

(I'm paranoid about how Markdown renders, sometimes without the space it might not render as a list.)

 * [Run mlpack bindings on a Raspberry Pi](crosscompile_armv7.md)
 * [Set up cross-compilation toolchain for mlpack](supported_boards.md)

> 

Reviewer: ```suggestion
```

### Cloning mlpack example respository

mlpack has an [example repository](https://github.com/mlpack/examples) that

> 

Reviewer: ```suggestion
```

shows a number of applications and use cases for mlpack, including embedded
deployment.  In this tutorial, we are interested in the `embedded/` directory,
which provides a CMake project template that compiles a random forest
application to embedded hardware.  This project template can be adapted to a new
project, or its pieces can be incorporated into an existing CMake project.

The first step is to clone the examples repository:

```sh
git clone git@github.com:mlpack/examples.git
```

Next, let's look at the `CMakeLists.txt` (e.g. the CMake configuration) in the
`embedded/crosscompile_random_forest/` directory.

### `CMakeLists.txt`

The first part of the code, printed below or

> [available here](https://github.com/mlpack/examples/blob/master/embedded/crosscompile_random_forest/CMakeLists.txt),

Reviewer: I guess we need to update that example with another PR?  It should be very simple I suppose.

Author: yes I can do it in another one and then come back to this one if necessary

Author: Yes

defines the project name and includes two useful CMake configuration files:
 * `mlpack.cmake`: finds mlpack's dependencies and download them if necessary.
 * `ConfigureCrossCompile.cmake`: set up CMake configuration for cross-compilation.
 * `crosscompile-toolchain.cmake`: invoke CMake crosscompilation infrastructure.
 * `crosscompile-arch-config.cmake`: add necessary flags depending on the
   architecture (optional).

Then we need to call `fetch_mlpack()` will download mlpack including all dependencies,

> cross-compile openblas and set up all the necessary parameters to find these dependencies.

Reviewer: ```suggestion
cross-compile OpenBLAS and set up all the necessary parameters to find these dependencies.
```

> Most of mlpack dependencies are header-only with the exception of Openblas,

Reviewer: ```suggestion
Most of mlpack's dependencies are header-only with the exception of OpenBLAS;
```

thus this is expected to be a quick step.

`fetch_mlpack()` will detect if cross compilation is necessary or not depending
on the command that is executed when running cmake. Based on this, it will
compile OpenBLAS.

```cmake
cmake_minimum_required(VERSION 3.11)
project(main)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake")
include(CMake/Autodownload.cmake)

> include(CMake/ConfigureCrossCompile.cmake)

Reviewer: I think these can be updated; at least, Autodownload.cmake isn't needed anymore.

> fetch_mlpack()

Reviewer: ```suggestion
// Download all of mlpack's dependencies and cross-compile OpenBLAS.
fetch_mlpack()
```
(just in case a user didn't read the text above, this comment gets the idea across :))

> 

Reviewer: ```suggestion
```

```

#### Setting up include directories and source files

> The last part of our CMake file consists of merging all of the above components

Reviewer: ```suggestion
The last part of our `CMakeFiles.txt` consists of merging all of the components above
```
Tiny grammatical issue; it is very pedantic :)

together. First we are going to find the OpenMP package if the user defined the

> variable above, then we will start including the directories of our

Reviewer: Is the text about OpenMP still necessary? I don't see anything about it in the code snippet below.

program. You will need to do the same for your software if you are trying to

> integrate this example into an existing development, this should be done by

Reviewer: ```suggestion
integrate this example into an existing codebase; this should be done by
```

defining an include variable using `set` directive or add them directly in
`target_include_directories`.
Regarding the source code, a similar process by either adding then to
`add_executables` or by appending the source files to `SOURCES_FILES` variable. 

Finally do not forget to add any external library that you need to link against
in `target_link_libraries`.

```cmake
include_directories(BEFORE ${MLPACK_INCLUDE_DIRS})

> include_directories(BEFORE ${CMAKE_CURRENT_SOURCE_DIR}/src/)

Reviewer: I'm not sure this `${CMAKE_CURRENT_SOURCE_DIR}/src/` include should be necessary for the example, especially if we are setting the target sources directly in `SOURCES_FILES`.

## User includes go here

# Finally, add any cross-compilation support libraries (they may need to come
# last).  If we are not cross-compiling, no changes will happen here.

> set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${CROSS_COMPILE_SUPPORT_LIBRARIES})

Reviewer: Do you mean that `CROSS_COMPILE_SUPPORT_LIBRARIES` is meant to be set by the user, or is that something that is done by `ConfigureCrossCompile.cmake`?

Author: usually, it is done by `ConfigureCrossCompile.cmake`

> ## Add your source files to SOURCES_FILES list

Reviewer: It might be clearer to explicitly list the sources here:

```
set(SOURCE_FILES
  main.cpp
)
```

add_executable(RandomForest main.cpp ${SOURCES_FILES})

> target_sources(RandomForest PRIVATE ${SOURCE_FILES})

Reviewer: Are both of these lines necessary?  I think you only need `add_executable()`; correct me if I'm wrong.

> ## Do not forget to add the include variables in here.
> 
> target_include_directories(RandomForest PRIVATE
>   ${MLPACK_INCLUDE_DIRS}
>   /path/to/your/own/include
> )

Reviewer: ```suggestion
# If needed, add any additional include directories here.
target_include_directories(RandomForest PRIVATE
  ${MLPACK_INCLUDE_DIRS}
  # Add more include directories here...
)
```

> ## Do not forget to add libraries that you are linking against in here.
> 
> target_link_libraries(RandomForest PRIVATE -static
>   ${MLPACK_LIBRARIES}
> )

Reviewer: ```suggestion
# If your application needs to link against more than just mlpack's
# dependencies, be sure to list them here.
target_link_libraries(RandomForest PRIVATE -static
  ${MLPACK_LIBRARIES}
  # List additional dependencies to link against here.
)
```

```

> #### Optimization and crosscompilation

Reviewer: ```suggestion
#### Optimization and cross-compilation
```

> If you are interested in adding specific compiler flags to optimize operations
> on your hardware. Please feel free to look at mlpack
> `CMake/crosscompile-arch-flags.cmake` you can either integrate the variable by
> copy and paste them directly to this CMake file or you can
> copy the entire `crosscompile-arch-config.cmake` file from `mlpack/CMake` and included
> it close to `Autodownload.cmake` at start of this tutorial but in this case you need
> to specify the `ARCH_NAME` variable as we did in the this [tutorial](crosscompile_armv7.md).

Reviewer: ```suggestion
If you are interested in adding specific compiler flags to optimize operations
on your hardware, you can either set `CMAKE_CXX_FLAGS` manually, or look at the
copy of `CMake/crosscompile-arch-flags.cmake` in your project, find the
appropriate `ARCH_NAME` section, and add the new compilation flags to that file.
```
Hopefully I did not change the meaning of this too much; I tried to just clean it up and shorten it as much as possible.

Once you have added all the source files and the headers for your applications,
you can create your own build directory and build the software using `cmake`,
and then `make`. Your cmake command should be similar to the following:

```sh
cmake \

>     -DBUILD_TESTS=ON \

Reviewer: ```suggestion
```
No tests needed in the example application, I think?

Author: nope, we can set this one to off

    -ARCH_NAME=(Check below) \
    -DCMAKE_CROSSCOMPILING=ON \
    -DCMAKE_TOOLCHAIN_FILE=../CMake/crosscompile-toolchain.cmake \
    -DTOOLCHAIN_PREFIX=(Check below) \
    -DCMAKE_SYSROOT=(Check below) \
```

> In order to fill the `TOOLCHAIN_PREFIX` or the `CMAKE_SYSROOT`, or if
> you are interested in a different compiler toolchain please refer to the
> following [table](supported_boards.md).

Reviewer: ```suggestion
In order to fill the `TOOLCHAIN_PREFIX` or the `CMAKE_SYSROOT`, use
[this table](supported_boards.md).

If your preferred architecture is missing, or if the table needs an update,
please submit a PR to the repository and help us keep it up to date!
```

I dunno, maybe it is helpful to request that users help keep things up to date.  Let me know what you think :smile:

### `CMakeLists.txt`

The first part of the code, printed below or
[available here](https://github.com/mlpack/examples/blob/master/embedded/crosscompile_random_forest/CMakeLists.txt),
defines the project name and includes two useful CMake configuration files:
 * `Autodownload.cmake`: downloads mlpack's dependencies from defined locations
 * `ConfigureCrossCompile.cmake`: set up CMake configuration for cross-compilation
 
The next steps set the required C++ standard to C++17 (which is the minimum required
version for mlpack), and enable OpenMP for parallelism.

```cmake
cmake_minimum_required(VERSION 3.11)
project(main)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake")
include(CMake/Autodownload.cmake)
include(CMake/ConfigureCrossCompile.cmake)

option(USE_OPENMP "If available, use OpenMP for parallelization." ON)
# Set required standard to C++17.
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

# If we're using gcc, then we need to link against pthreads to use std::thread,
# which we do in the tests.
if (CMAKE_COMPILER_IS_GNUCC)
  find_package(Threads)
  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})

> endif()

Reviewer: So I think that you haven't rewritten this part yet, but I really feel pretty strongly that for a non-cross-compiled example, the following CMake code should work:

```
cmake_minimum_required(VERSION 3.11)
project(my_mlpack_code C++)

# Findmlpack.cmake needs to be available, so maybe a `set(CMAKE_MODULE_PATH ...)` line is needed
find_package(mlpack 4.5.2 REQUIRED)

include_directories(${MLPACK_INCLUDE_DIRS})
add_executable(my_mlpack_program my_mlpack_program.cpp)
target_link_libraries(my_mlpack_program ${MLPACK_LIBRARIES})
```

and that should be all that's needed.  I feel pretty strongly about that because a user who wants to include mlpack in their project should have to do as little as possible for the "simple" case.

Now you are concerned with cross-compilation where you want to manually cross-compile OpenBLAS, but I think this can fit too with a slight modification:

```
cmake_minimum_required(VERSION 3.11)
project(my_mlpack_code C++)

# Set cross-compilation options.
set(ARCH_NAME arm64) # or something like this, the user modifies it
# The autodownloader can live in this file
include(CMake/mlpackDependencyDownloader.cmake)
# This will use ARCH_NAME internally, then download and compile OpenBLAS
mlpack_download_dependencies()
# Now, variables used by the Findmlpack.cmake script are set, like `BLAS_LIBRARIES` or similar,
# so that Findmlpack.cmake finds the ones that we just downloaded.  In fact I think you can
# just set ARMADILLO_INCLUDE_DIR, ARMADILLO_LIBRARIES, CEREAL_INCLUDE_DIR, and
# ENSMALLEN_INCLUDE_DIR directly, and then Findmlpack.cmake will use those directly
# instead of trying to find other versions.

# Findmlpack.cmake needs to be available, so maybe a `set(CMAKE_MODULE_PATH ...)` line is needed
# Note that this sets MLPACK_INCLUDE_DIRS to *all* include directories needed for
# mlpack *and* its dependencies, and MLPACK_LIBRARIES contains all runtime libraries
# that needs to be linked for mlpack's dependencies (generally just Armadillo, but if we aren't
# using the wrapper it could be a bunch).
find_package(mlpack 4.5.2 REQUIRED)

include_directories(${MLPACK_INCLUDE_DIRS})
add_executable(my_mlpack_program my_mlpack_program.cpp)
target_link_libraries(my_mlpack_program ${MLPACK_LIBRARIES})
```

So basically my priority is separating the "regular" use of mlpack inside of CMake, which should be a one-liner, from the cross-compilation use where we download and set things up, which should be a two-liner.  I think that this is a relatively small change from what you have now, the main difference being that right now `Dependencies.cmake` both downloads and finds the dependencies, but my suggestion is that the 'finding dependency' part should also be a part of `Findmlpack.cmake`.

I know I wrote a lot (not unusual for me I guess :smile:), let me know if I can clarify anything of what I meant or if I overlooked something.
